### PR TITLE
fix setting of coreos userdata on vSphere

### PR DESCRIPTION
**Pull request by mrIncompetent**
**What this PR does / why we need it**:
Finally fixed the e2e tests.
Was picked from https://github.com/kubermatic/machine-controller/pull/214

### CODE REVIEW

File: pkg/cloudprovider/provider/vsphere/helper.go [hunk 1]
```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
...
-			{
-				Info:            &types.VAppPropertyInfo{Key: 7, Id: "guestinfo.coreos.config.data.encoding", Value: "base64"},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-			{
-				Info:            &types.VAppPropertyInfo{Key: 4, Id: "guestinfo.coreos.config.data", Value: userdataBase64},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-		}}
+		// The properties describing userdata will already exist in the CoreOS VM template.
```
[the remark this answers is not in the document]
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: kudos to @kdomanski

File: pkg/cloudprovider/provider/vsphere/helper.go [hunk 2]
```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
...
+		// In order to overwrite them, we need to specify their numeric Key values,
+		// which we'll extract from that template.
+		var mvm mo.VirtualMachine
+		err = templateVm.Properties(ctx, templateVm.Reference(), []string{"config", "config.vAppConfig", "config.vAppConfig.property"}, &mvm)
+		if err != nil {
+			return err
+		}
+
+		var property []types.VAppPropertySpec
+		info := mvm.Config.VAppConfig.GetVmConfigInfo()
```
Comment (p0lyn0mial) [inline pkg/cloudprovider/provider/vsphere/helper.go]: doesn't work in golang ?

`properties := mvm.Config.VAppConfig.GetVmConfigInfo().Property`

File: pkg/cloudprovider/provider/vsphere/helper.go [hunk 3]
```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
...
+		properties := info.Property
+		for _, item := range properties {
+			switch item.Id {
+			case "guestinfo.coreos.config.data":
+				property = append(property, types.VAppPropertySpec{
+					ArrayUpdateSpec: types.ArrayUpdateSpec{
```
Comment (p0lyn0mial) [inline pkg/cloudprovider/provider/vsphere/helper.go]: just wanted to see if this doesn't have to be a pointer. the source code looks really ugly.

```
type ArrayUpdateSpec struct {

   DynamicData
    
   Operation ArrayUpdateOperation `xml:"operation"`
   RemoveKey AnyType              `xml:"removeKey,omitempty,typeattr"`
   }
    
   func init() {
   t["ArrayUpdateSpec"] = reflect.TypeOf((*ArrayUpdateSpec)(nil)).Elem()
   }
```

File: pkg/cloudprovider/provider/vsphere/helper.go (hunk 2 above)
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: It does but i encountered at some point a nil-pointer exception. I was not able to reproduce it but splitting the one line makes it easier to debug later on

File: pkg/cloudprovider/provider/vsphere/helper.go [hunk 4]
```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
...
-			{
-				Info:            &types.VAppPropertyInfo{Key: 4, Id: "guestinfo.coreos.config.data", Value: userdataBase64},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-		}}
+		// The properties describing userdata will already exist in the CoreOS VM template.
+		// In order to overwrite them, we need to specify their numeric Key values,
+		// which we'll extract from that template.
+		var mvm mo.VirtualMachine
+		err = templateVm.Properties(ctx, templateVm.Reference(), []string{"config", "config.vAppConfig", "config.vAppConfig.property"}, &mvm)
```
[the remark this answers is not in the document]
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: I'm not entirely sure if they'll always exist. They definitely do on the official coreos-template.

File: pkg/cloudprovider/provider/vsphere/helper.go [hunk 5]
```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
...
+						Key:   item.Key,
+						Id:    item.Id,
+						Value: "base64",
+					},
+				})
+				break
+			}
+		}
+
+		vAppAconfig = &types.VmConfigSpec{Property: property}
```
Comment (p0lyn0mial) [inline pkg/cloudprovider/provider/vsphere/helper.go]: oh, we are taking only the first encountered value.

**Comment by p0lyn0mial** [comment]
Do we have to export `CreateLinkClonedVm` method ?

File: pkg/cloudprovider/provider/vsphere/helper.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/cloudprovider/provider/vsphere/helper.go]: oh, should we then add the following check ?
`if info != nil {...}`

**Comment by p0lyn0mial** [comment]
it seems like the API returns only pointers, checking every returned value against nil seems counter productive. It seems like the authors of the `govoc` don't evaluate the pointers 

https://github.com/vmware/govmomi/blob/master/govc/vapp/destroy.go

**Comment by kdomanski** [comment]
I could have just rebased the old PR against master.

**Comment by kdomanski** [comment]
Nah, just pointing out for the future.

File: pkg/cloudprovider/provider/vsphere/helper.go (hunk 2 above)
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: Was actually just one thing i needed to check.
Now we return an meaningful error

File: pkg/cloudprovider/provider/vsphere/helper.go (hunk 3 above)
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: Yeah, the vmware sdk is quite special...

File: pkg/cloudprovider/provider/vsphere/helper.go (hunk 5 above)
Comment (mrIncompetent) [reply pkg/cloudprovider/provider/vsphere/helper.go]: `property` was actually a slice. But as the naming of the vmware type is fairly bad, it's super confusing.
I renamed the var to make it more clear

**Comment by mrIncompetent** [comment]
> Do we have to export CreateLinkClonedVm method ?

Nope, changed

**Comment by mrIncompetent** [comment]
PTAL @p0lyn0mial

**Pull request merged**